File: frontend/src/queries/analytics.ts
/**
 * Analytics query hooks. All endpoints share the same trade-filter shape so
 * the user can carry filter chips from the trade log into analytics.
 */

import { useMutation, useQuery } from '@tanstack/react-query';
import { api } from '@/lib/api';

// --- Filters shared with /trades -----------------------------------------

export interface AnalyticsFilters {
  strategy_id?: string;
  league?: string;
  status?: 'OPEN' | 'CLOSED' | 'VOID';
  outcome_label?: string;
  pnl_mode?: 'AUTO' | 'MANUAL' | 'CASHOUT_ODDS';
  date_from?: string;
  date_to?: string;
  pnl_min?: string;
  pnl_max?: string;
  tag?: string[];
  q?: string;
}

function asParams(f: AnalyticsFilters): Record<string, string | number | string[] | undefined> {
  const params: Record<string, string | number | string[] | undefined> = {};
  for (const [k, v] of Object.entries(f)) {
    if (v === undefined || v === null || v === '') continue;
    if (Array.isArray(v) && v.length === 0) continue;
    params[k] = v as string | number | string[];
  }
  return params;
}

// --- Summary --------------------------------------------------------------

export interface AnalyticsSummary {
  n_trades: number;
  total_pnl_eur: string;
  total_stake_eur: string;
  roi_pct: string;
  win_rate_pct: string;
  sharpe: string;
  max_drawdown_pct: string;
  max_drawdown_eur: string;
}

export function useSummary(f: AnalyticsFilters) {
  return useQuery({
    queryKey: ['analytics', 'summary', asParams(f)],
    queryFn: () => api.get<AnalyticsSummary>('/analytics/summary', asParams(f)),
  });
}

// --- Breakdowns -----------------------------------------------------------

export interface BreakdownRow {
  key: string;
  n_trades: number;
  total_pnl_eur: string;
  total_stake_eur: string;
  roi_pct: string;
  win_rate_pct: string;
}

export function useByStrategy(f: AnalyticsFilters) {
  return useQuery({
    queryKey: ['analytics', 'by-strategy', asParams(f)],
    queryFn: () => api.get<BreakdownRow[]>('/analytics/by-strategy', asParams(f)),
  });
}

export function useByLeague(f: AnalyticsFilters) {
  return useQuery({
    queryKey: ['analytics', 'by-league', asParams(f)],
    queryFn: () => api.get<BreakdownRow[]>('/analytics/by-league', asParams(f)),
  });
}

export function useByOutcome(f: AnalyticsFilters) {
  return useQuery({
    queryKey: ['analytics', 'by-outcome', asParams(f)],
    queryFn: () => api.get<BreakdownRow[]>('/analytics/by-outcome', asParams(f)),
  });
}

// --- Rolling -------------------------------------------------------------

export interface RollingPoint {
  idx: number;
  roi_pct: string;
  win_rate_pct: string;
}

export function useRolling(f: AnalyticsFilters, window = 20) {
  return useQuery({
    queryKey: ['analytics', 'rolling', window, asParams(f)],
    queryFn: () => api.get<RollingPoint[]>('/analytics/rolling', { ...asParams(f), window }),
  });
}

// --- Drawdown ------------------------------------------------------------

export interface DrawdownPoint {
  closed_at: string;
  cum_pnl_eur: string;
  underwater_eur: string;
  underwater_pct: string;
}

export interface DrawdownSeries {
  points: DrawdownPoint[];
  max_drawdown_pct: string;
  max_drawdown_eur: string;
  max_dd_started_at: string | null;
  max_dd_ended_at: string | null;
}

export function useDrawdown(f: AnalyticsFilters) {
  return useQuery({
    queryKey: ['analytics', 'drawdown', asParams(f)],
    queryFn: () => api.get<DrawdownSeries>('/analytics/drawdown', asParams(f)),
  });
}

// --- Calendar ------------------------------------------------------------

export interface CalendarCell {
  day_of_week: number;
  hour: number;
  n_trades: number;
  pnl_eur: string;
}

export function useCalendar(f: AnalyticsFilters) {
  return useQuery({
    queryKey: ['analytics', 'calendar', asParams(f)],
    queryFn: () => api.get<{ cells: CalendarCell[] }>('/analytics/calendar', asParams(f)),
  });
}

// --- Monte Carlo (mutation: triggered on demand) -------------------------

export interface MonteCarloRequest {
  starting_bankroll: string;
  n_simulations: number;
  horizon_trades: number;
  ruin_threshold_pct: string;
  n_buckets?: number;
  seed?: number;
}

export interface DistributionBucket {
  bucket_low: string;
  bucket_high: string;
  count: number;
}

export interface MonteCarloResponse {
  risk_of_ruin_pct: string;
  p10_ending_bankroll: string;
  p50_ending_bankroll: string;
  p90_ending_bankroll: string;
  mean_ending_bankroll: string;
  min_ending_bankroll: string;
  max_ending_bankroll: string;
  distribution: DistributionBucket[];
  n_simulations: number;
  horizon_trades: number;
  n_historical_pnls: number;
}

export function useMonteCarlo(f: AnalyticsFilters) {
  return useMutation({
    mutationFn: (body: MonteCarloRequest) =>
      api.post<MonteCarloResponse>('/analytics/monte-carlo', body, asParams(f)),
  });
}

// --- WhatIf (mutation: keystroke-debounced from the widget) --------------

export interface WhatIfCashoutRequest {
  stake_total: string;
  avg_odds: string;
  cashout_odds: string;
  position_side: 'back' | 'lay';
  commission_pct: string;
  market_type?: 'exchange' | 'classic';
}

export interface WhatIfCashoutResponse {
  locked_in_pnl_eur: string;
  breakeven_cashout_odds: string | null;
  pct_of_max_win: string;
  formula_text: string;
}

export function useWhatIfCashout() {
  return useMutation({
    mutationFn: (body: WhatIfCashoutRequest) =>
      api.post<WhatIfCashoutResponse>('/analytics/whatif-cashout', body),
  });
}

File: backend/app/schemas/analytics.py
"""Analytics response + request schemas."""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field

from app.models import MarketType


# --- Summary ---------------------------------------------------------------


class AnalyticsSummary(BaseModel):
    n_trades: int
    total_pnl_eur: Decimal
    total_stake_eur: Decimal
    roi_pct: Decimal
    win_rate_pct: Decimal
    sharpe: Decimal
    max_drawdown_pct: Decimal
    max_drawdown_eur: Decimal


# --- Breakdowns ------------------------------------------------------------


class BreakdownRow(BaseModel):
    key: str
    n_trades: int
    total_pnl_eur: Decimal
    total_stake_eur: Decimal
    roi_pct: Decimal
    win_rate_pct: Decimal


# --- Rolling ---------------------------------------------------------------


class RollingPoint(BaseModel):
    idx: int
    roi_pct: Decimal
    win_rate_pct: Decimal


# --- Drawdown --------------------------------------------------------------


class DrawdownPoint(BaseModel):
    closed_at: datetime
    cum_pnl_eur: Decimal
    underwater_eur: Decimal
    underwater_pct: Decimal


class DrawdownSeries(BaseModel):
    points: list[DrawdownPoint]
    max_drawdown_pct: Decimal
    max_drawdown_eur: Decimal
    max_dd_started_at: datetime | None
    max_dd_ended_at: datetime | None


# --- Calendar --------------------------------------------------------------


class CalendarCell(BaseModel):
    day_of_week: int
    hour: int
    n_trades: int
    pnl_eur: Decimal


class CalendarGrid(BaseModel):
    cells: list[CalendarCell]


# --- Monte Carlo -----------------------------------------------------------


class MonteCarloRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    starting_bankroll: Annotated[Decimal, Field(gt=Decimal("0"))]
    n_simulations: Annotated[int, Field(ge=1, le=100_000)] = 10_000
    horizon_trades: Annotated[int, Field(ge=1, le=1_000)] = 100
    ruin_threshold_pct: Annotated[Decimal, Field(ge=Decimal("0"), le=Decimal("100"))] = (
        Decimal("50.0")
    )
    n_buckets: Annotated[int, Field(ge=2, le=200)] = 20
    seed: int | None = None


class DistributionBucket(BaseModel):
    bucket_low: Decimal
    bucket_high: Decimal
    count: int


class MonteCarloResponse(BaseModel):
    risk_of_ruin_pct: Decimal
    p10_ending_bankroll: Decimal
    p50_ending_bankroll: Decimal
    p90_ending_bankroll: Decimal
    mean_ending_bankroll: Decimal
    min_ending_bankroll: Decimal
    max_ending_bankroll: Decimal
    distribution: list[DistributionBucket]
    n_simulations: int
    horizon_trades: int
    n_historical_pnls: int


# --- WhatIf cash-out -------------------------------------------------------


class WhatIfCashoutRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    stake_total: Annotated[Decimal, Field(ge=Decimal("0"))]
    avg_odds: Annotated[Decimal, Field(ge=Decimal("1.01"))]
    cashout_odds: Annotated[Decimal, Field(ge=Decimal("0"))]
    position_side: Literal["back", "lay"]
    commission_pct: Annotated[
        Decimal, Field(ge=Decimal("0"), le=Decimal("100"))
    ] = Decimal("5.0")
    market_type: MarketType = MarketType.exchange


class WhatIfCashoutResponse(BaseModel):
    locked_in_pnl_eur: Decimal
    breakeven_cashout_odds: Decimal | None
    pct_of_max_win: Decimal
    formula_text: str
